src/cri_attributes.rs

```diff
@@ -99,12 +122,43 @@ pub(crate) mod parser {
             .map(|(i, extensions)| (i, ExtensionRequest { extensions }))
     }
 
-    fn parse_extension_request_ext(i: &[u8]) -> X509Result<ParsedCriAttribute> {
+    fn parse_extension_request_attr(i: &[u8]) -> X509Result<ParsedCriAttribute> {
         map(
             parse_extension_request,
             ParsedCriAttribute::ExtensionRequest,
         )(i)
     }
+
+    pub(super) fn parse_challenge_password(i: &[u8]) -> X509Result<ChallengePassword> {
+        let (rem, obj) = match alt((
+            parse_der_utf8string,
+            parse_der_printablestring,
+            parse_der_numericstring,
+            parse_der_bmpstring,
+            visiblestring,
+            parse_der_generalstring,
+            parse_der_objectdescriptor,
+            parse_der_graphicstring,
+            parse_der_t61string,
+            parse_der_videotexstring,
+            parse_der_ia5string,
+        ))(i)
```

Comment:
Hum, this is not very clear to me (and the new `alt` parser is quite complex - and possibly slow).
I have found `challengePassword` in section 5.4.1 of RFC 2985, defined as `DirectoryString`.
Now, `DirectoryString can be found in RFC 5280:
```
   DirectoryString ::= CHOICE {
         teletexString           TeletexString (SIZE (1..MAX)),
         printableString         PrintableString (SIZE (1..MAX)),
         universalString         UniversalString (SIZE (1..MAX)),
         utf8String              UTF8String (SIZE (1..MAX)),
         bmpString               BMPString (SIZE (1..MAX)) }
```
I don't really see why `ObjectDescriptor`, `NumericString`, etc are listed here.
So my guess is that you probably do not need all these types, but only those from the list. 

You should probably keep the ASN.1 definition in a comment for the function - this helps a lot when trying to interpret the code.

Also, you should order them by "probability" of being found - the `alt` parser will try them one at a time, so the order is important and you should keep `PrintableString` and `UTF8String` parsers at the beginning.

Reply:
Correct. I missed the last 3 words in the RFC:
```
   ChallengePassword attribute values generated in accordance with this
   version of this document SHOULD use the PrintableString encoding
   whenever possible.  If internationalization issues make this
   impossible, the UTF8String alternative SHOULD be used.  PKCS #9-
   attribute processing systems MUST be able to recognize and process
   all string types in DirectoryString values.
```
Should be fixed and I kept `UTF8String` and `PrintableString` in the pole position.